src/lib/device-proxy.ts

```diff
@@ -243,26 +244,33 @@ export function requestDevices({
 						const vpnIp = device.is_managed_by__service_instance[0].ip_address;
 						const deviceUrl = `http://${device.uuid}.balena:${device.api_port ||
 							80}${url}?apikey=${device.api_secret}`;
-						promises.push(
-							requestAsync({
-								uri: deviceUrl,
-								json: data,
-								proxy: `http://resin_api:${API_VPN_SERVICE_API_KEY}@${vpnIp}:3128`,
-								tunnel: true,
-								method: method,
-								timeout: DEVICE_REQUEST_TIMEOUT,
-							}),
-						);
-						// We add a delay between each notification so that we do not in essence trigger a DDOS from resin
-						// devices against us, but we do not wait for completion of individual requests because doing so
-						// could cause a terrible UX if we have a device time out, as that would block all the subsequent
-						// notifications
+						let p = requestAsync({
+							uri: deviceUrl,
+							json: data,
+							proxy: `http://resin_api:${API_VPN_SERVICE_API_KEY}@${vpnIp}:3128`,
+							tunnel: true,
+							method: method,
+							timeout: DEVICE_REQUEST_TIMEOUT,
+						});
+						if (!wait) {
+							// this force-cast is super ugly but harmless because clearly noone
+							// cares about the return value of the promise (since wait == false)
+							// so we just need to satisfy the compiler
+							p = p.catchReturn((null as any) as RequestResponse);
```

Comment:
For some reason I can't understand, the catch handler needs to be on the request promise itself -- if I do `Promise.all(promises).catchReturn(null)` below or `waitPromise.catchReturn()` further down, the exception still propagates. I believe it may have something to do with the promisification and the fact that the error is emit from an event handler, though I do feel that's surprising.

Reply:
It'll be due to when the error handler is added, if you made the request fail after 10s instead then the handler would be added before the rejection actually happens and so should be fine